### PR TITLE
Enable Modularisation

### PR DESCRIPTION
**Pull request by fluffyemily**
Extract Store, Logins, Utils, Categories and Items into separate crates.

- store
- logins
- list
- ffi-utils

Separate out read connection function from utils to store
Make store a dependency for logins and list rather than the other way around
Rename library from greetings to toodle
Provide Toodle object that holds handles to logins, store and list and provides FFI interface for accessing.
Update Xcode project to store and call the new structures.

#5

### CODE REVIEW

File: complex objects/cargo/list/src/categories.rs [hunk 1]
```diff
@@ -15,40 +15,36 @@ use std::os::raw::{
...
+    store: Arc<Store>,
 }
 
 impl CategoryManager {
-    pub fn new(uri: String, conn: Arc<Mutex<Connection>>) -> CategoryManager {
-        CategoryManager {
-            conn: conn,
-            uri: uri
-        }
+    pub fn new(store: Arc<Store>) -> CategoryManager {
```
Comment (rnewman) [inline complex objects/cargo/list/src/categories.rs]: This commit is a bit more than a refactor, no? Might be worth flattening and/or re-splitting.

File: complex objects/cargo/Cargo.toml [hunk 2]
```diff
@@ -2,6 +2,7 @@
 name = "toodle"
 version = "0.1.0"
 authors = ["Emily Toop <etoop@mozilla.com>"]
+description = "Library for providing To Do List data cross platform"
```
Comment (rnewman) [inline complex objects/cargo/Cargo.toml]: "across platforms" or "Cross-platform library"

File: complex objects/cargo/list/src/lib.rs [hunk 3]
```diff
@@ -195,14 +197,18 @@ impl ListManager {
                 match item_iter.next() {
                     Some(result) => {
                         match result {
-                            Ok(item) => Some(item),
+                            Ok(item) => {
+                                println!("Returning created item {:?}", item);
+                                Some(item)
```
Comment (rnewman) [inline complex objects/cargo/list/src/lib.rs]: FYI, `Result::ok()` turns a `Result` into an `Option`. 

Digression into ways to handle this kind of thing: we added some [utilities in Mentat](https://github.com/mozilla/mentat/blob/20aa11dcbdabd03b0274b849f640caff84b46882/core/src/util.rs#L11) to chain side-effects off both `Result` and `Option`, so you could replace this whole `match` block with something flatter, like:

```
item_iter.next()
         .when_ok((item) => { println!("Returning created item {:?}", item); })
         .when_err((e) => { println!("Failed to fetch item {:?}", e); })
         .ok()
```

and then just kill those `when` lines when you no longer want the debugging.

File: complex objects/cargo/src/lib.rs [hunk 4]
```diff
@@ -55,13 +55,13 @@ pub unsafe extern "C" fn toodle_destroy(toodle: *mut Toodle) {
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn toodle_logins(toodle: *mut Toodle) -> *mut Arc<LoginManager> {
+pub unsafe extern "C" fn toodle_logins(toodle: *mut Arc<Toodle>) -> *mut Arc<LoginManager> {
```
Comment (rnewman) [inline complex objects/cargo/src/lib.rs]: How does refcounting work across the FFI boundary? My understanding was that it doesn't do anything special, so you're reliant on that `Arc` being held somewhere else, no?

Comment (fluffyemily) [reply complex objects/cargo/src/lib.rs]: Yeah, I think I was just Arc happy at that point.

**Comment by fluffyemily** [comment]
@rnewman I refactored and made those changes you requested.

I couldn't get `ResultEffect` to work as I needed - I think I need to amend the function to take an arg so that I can get access to the value of the `Result` to do stuff, but I figured that could be done in a separate PR.

**Pull request merged**